src/ucp/wireup/wireup_cm.c

```diff
@@ -443,50 +448,58 @@ static unsigned ucp_cm_client_uct_connect_progress(void *arg)
+    for (i = 0; i < ucs_static_array_size(ep_init_flags_variants); ++i) {
+        ep_init_flags = ep_init_flags_variants[i];
 
-    status = ucp_worker_get_ep_config(worker, &key, ep_init_flags,
-                                      &ep->cfg_index);
-    if (status != UCS_OK) {
-        goto err;
-    }
+        /* Cleanup the previously created UCP EP. The one that was created on
+         * the previous call to this client's resolve_cb */
+        ucp_wireup_cm_ep_cleanup(ep, &tmp_pending_queue);
 
-    ep->am_lane = key.am_lane;
+        status = ucp_cm_ep_client_initial_config_get(ep, ep_init_flags,
+                                        &cm_wireup_ep->cm_resolve_tl_bitmap,
+                                        &key);
+        if (status != UCS_OK) {
+            goto try_fallback;
+        }
 
-    status = ucp_cm_ep_init_lanes(ep, &tl_bitmap);
-    if (status != UCS_OK) {
-        goto err;
-    }
+        status = ucp_worker_get_ep_config(worker, &key, ep_init_flags,
+                                          &ep->cfg_index);
+        if (status != UCS_OK) {
+            goto err;
+        }
+
+        ep->am_lane = key.am_lane;
 
-    /* Replay pending requests from the tmp_pending_queue */
-    ucp_wireup_replay_pending_requests(ep, &tmp_pending_queue);
-
-    can_fallback = (ep_init_flags != UCP_EP_INIT_CREATE_AM_LANE_ONLY) ||
-                   (ucp_cm_client_get_next_cm_idx(ep) != UCP_NULL_RESOURCE);
-    status       = ucp_cm_ep_priv_data_pack(
-                      ep, &tl_bitmap, can_fallback,
-                      context->config.ext.sa_client_min_hdr_version,
-                      &priv_data, &priv_data_length, ep_init_flags);
-    if (status == UCS_ERR_BUFFER_TOO_SMALL) {
-        if (ep_init_flags != UCP_EP_INIT_CREATE_AM_LANE_ONLY) {
-            /* Create endpoint configuration with AM lane only to fit CM private
-             * data length limit */
-            ep_init_flags = UCP_EP_INIT_CREATE_AM_LANE_ONLY;
-            goto initial_config_retry;
+        status = ucp_cm_ep_init_lanes(ep, &tl_bitmap);
+        if (status != UCS_OK) {
+            goto err;
         }
 
+        /* Replay pending requests from the tmp_pending_queue */
+        ucp_wireup_replay_pending_requests(ep, &tmp_pending_queue);
+
+        can_fallback =
+                (i < (ucs_static_array_size(ep_init_flags_variants) - 1)) ||
+                (ucp_cm_client_get_next_cm_idx(ep) != UCP_NULL_RESOURCE);
+        status       = ucp_cm_ep_priv_data_pack(
+                          ep, &tl_bitmap, can_fallback,
+                          context->config.ext.sa_client_min_hdr_version,
+                          &priv_data, &priv_data_length, ep_init_flags);
+        if (status == UCS_ERR_BUFFER_TOO_SMALL) {
+            continue;
+        } else if (status != UCS_OK) {
+            goto err;
+        }
+
+        break;
+    }
+
+    if ((status == UCS_ERR_BUFFER_TOO_SMALL) &&
+        (i == ucs_static_array_size(ep_init_flags_variants))) {
```

Comment:
why need to check `i`?

Reply:
yeah, maybe just check if (status != UCS_OK)

Reply:
done